lib/util/source-code.js

```diff
@@ -111,9 +111,12 @@ class SourceCode extends TokenStore {
      * Represents parsed source code.
      * @param {string} text - The source code text.
      * @param {ASTNode} ast - The Program node of the AST representing the code. This AST should be created from the text that BOM was stripped.
+     * @param {Object|null} parserServices - The parser srevices.
+     * @param {ScopeManager|null} scopeManager - The scope of this source code.
+     * @param {Object|null} visitorKeys - The visitor keys to traverse AST.
      * @constructor
      */
-    constructor(text, ast) {
+    constructor(text, ast, parserServices, scopeManager, visitorKeys) {
```

Comment:
Those are optional because custom parsers not generate those always. However, if custom parsers generate those, those parameters are required. So I think those parameters are the same level as `text` and `ast`. Maybe one object parameter is better?

```js
new SourceCode({
    text,
    ast,
    parserServices,
    scopeManager,
    visitorKeys
});
```